Use NAPI to improve perfomance

furstenheim opened the pull request:
Hi @schnerd,
I've improved the performance for node using N-API. The pr is a bit long so there will be things to improve for sure. For example, one thing that I could not find was how to test for N-API properly. Right now it is checking Node version >= 8.

Basically what it does is that if the process is running in node with version greater or equal to 8, and data is big then it will use native add on to improve the performance.

This is js times:

    Benchmarking length 10: 1.982ms 
    Benchmarking length 100: 5.511ms
    Benchmarking length 1000: 14.654ms
    Benchmarking length 10000: 349.275ms
    Benchmarking length 100000: 32439.441ms

And this native times:
    
    Benchmarking length 10: 11.834ms
    Benchmarking length 100: 0.175ms
    Benchmarking length 1000: 5.621ms
    Benchmarking length 10000: 277.416ms
    Benchmarking length 100000: 22709.475ms

So something around the 50% improvement.

furstenheim committed:
Add benchmark

furstenheim committed:
Get typed array from input

furstenheim committed:
Sort input

furstenheim committed:
Add fill matrix method

furstenheim committed:
Fix compiling errors

furstenheim committed:
Initialize memory after malloc

furstenheim committed:
Return indices

furstenheim committed:
Fix algorithm

furstenheim committed:
Add triage file

furstenheim committed:
Fix errors in tests

furstenheim committed:
Fix triage

furstenheim committed:
Run tests with v9

furstenheim committed:
Split tests in modern and old

furstenheim committed:
Change tests depending on version

furstenheim committed:
Fix bash

furstenheim committed:
Add missing space

furstenheim committed:
Merge remote-tracking branch 'upstream/master'

furstenheim committed:
Lint tests

furstenheim committed:
Add semicolons

furstenheim committed:
Remove multi spaces

furstenheim committed:
Clean up benchmarking

schnerd commented:
Nice @furstenheim, will take a deeper look at this later today or tomorrow.

schnerd commented on package.json:
```diff
@@ -6,12 +6,19 @@
   "license": "MIT",
   "devDependencies": {
+    "bindings": "^1.3.0",
+    "d3": "^4.2.2",
```
Is d3 actually used anyhwere?

furstenheim replied:
No, missed that out when merging from upstream

schnerd commented on spec/CkmeansNodeModernSpec.js:
```diff
@@ -0,0 +1,18 @@
+      const length = parseInt(Math.random() * 10 + 300);
+      const data = [];
+      for (let j = 0; j < length; j++) {
+        data[j] = parseInt(Math.random() * 100);
```
Using random values in tests usually isn't ideal since results may not be reproducible across runs. Any way we can make this test deterministic?

schnerd commented on src/native/common.h:
```diff
@@ -0,0 +1,61 @@
+// Straight from node source code https://github.com/nodejs/node/blob/master/test/addons-napi/common.h
```
Forgive me, I'm not too familiar with NAPI–is it expected that all NAPI projects include a copy of this file? Seems like something the node runtime should provide.

furstenheim replied:
That was my initial thought, but I got it as not defined

schnerd commented on src/native/ckmeans.c:
```diff
@@ -0,0 +1,184 @@
+    }
+    return b;
+}
+// TODO
```
What's left TODO here?

furstenheim replied:
It was only during the write up of the c code

schnerd commented:
This causes errors when building the existing `dist/d3-scale-cluster.min.js` file:

```
$ npm run build

> d3-scale-cluster@1.2.0 build REDACTED/d3-scale-cluster
> webpack

Hash: 6a15fc6c6274a9b1d442
Version: webpack 1.15.0
Time: 88ms
                  Asset     Size  Chunks             Chunk Names
d3-scale-cluster.min.js  32.6 kB       0  [emitted]  main
   [0] ./src/index.js 2.33 kB {0} [built]
   [1] ./src/ckmeans-triage.js 499 bytes {0} [built]
   [3] ./src/ckmeans.js 8.88 kB {0} [built]
   [4] ./src/ckmeans-native.js 208 bytes {0} [built]
    + 6 hidden modules

WARNING in ./~/bindings/bindings.js
Critical dependencies:
81:22-40 the request of a dependency is an expression
81:43-53 the request of a dependency is an expression
 @ ./~/bindings/bindings.js 81:22-40 81:43-53

WARNING in ./~/bindings/package.json
Module parse failed: REDACTED/d3-scale-cluster/node_modules/bindings/package.json Unexpected token (2:9)
You may need an appropriate loader to handle this file type.
SyntaxError: Unexpected token (2:9)
    at Parser.pp$4.raise (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:2221:15)
    at Parser.pp.unexpected (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:603:10)
    at Parser.pp.semicolon (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:581:61)
    at Parser.pp$1.parseExpressionStatement (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:966:10)
    at Parser.pp$1.parseStatement (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:730:24)
    at Parser.pp$1.parseBlock (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:981:25)
    at Parser.pp$1.parseStatement (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:709:33)
    at Parser.pp$1.parseTopLevel (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:638:25)
    at Parser.parse (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:516:17)
    at Object.parse (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:3098:39)
    at Parser.parse (REDACTED/d3-scale-cluster/node_modules/webpack/lib/Parser.js:902:15)
    at NormalModule.<anonymous> (REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModule.js:104:16)
    at NormalModule.onModuleBuild (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:310:10)
    at nextLoader (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:275:25)
    at REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:259:5
    at Storage.provide (REDACTED/d3-scale-cluster/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:52:20)
    at CachedInputFileSystem.readFile (REDACTED/d3-scale-cluster/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:140:24)
    at NormalModule.onLoadPitchDone (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:255:7)
    at NormalModule.loadPitch (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:182:27)
    at NormalModule.doBuild (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:241:4)
    at NormalModule.build (REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModule.js:84:14)
    at Compilation.buildModule (REDACTED/d3-scale-cluster/node_modules/webpack/lib/Compilation.js:126:9)
    at REDACTED/d3-scale-cluster/node_modules/webpack/lib/Compilation.js:309:10
    at REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModuleFactory.js:58:13
    at NormalModuleFactory.applyPluginsAsyncWaterfall (REDACTED/d3-scale-cluster/node_modules/tapable/lib/Tapable.js:75:69)
    at onDoneResolving (REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModuleFactory.js:38:11)
    at onDoneResolving (REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModuleFactory.js:121:6)
    at REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModuleFactory.js:116:7
    at REDACTED/d3-scale-cluster/node_modules/async/lib/async.js:726:13
    at REDACTED/d3-scale-cluster/node_modules/async/lib/async.js:52:16
 @ ./~/bindings ^\.\/.*$

WARNING in ./~/bindings/README.md
Module parse failed: REDACTED/d3-scale-cluster/node_modules/bindings/README.md Unexpected token (2:3)
You may need an appropriate loader to handle this file type.
SyntaxError: Unexpected token (2:3)
    at Parser.pp$4.raise (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:2221:15)
    at Parser.pp.unexpected (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:603:10)
    at Parser.pp$3.parseExprAtom (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1822:12)
    at Parser.pp$3.parseExprSubscripts (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1715:21)
    at Parser.pp$3.parseMaybeUnary (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1692:19)
    at Parser.pp$3.parseExprOp (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1656:43)
    at Parser.pp$3.parseExprOp (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1658:21)
    at Parser.pp$3.parseExprOps (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1639:17)
    at Parser.pp$3.parseMaybeConditional (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1620:21)
    at Parser.pp$3.parseMaybeAssign (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1597:21)
    at Parser.pp$3.parseExpression (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:1573:21)
    at Parser.pp$1.parseStatement (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:727:47)
    at Parser.pp$1.parseTopLevel (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:638:25)
    at Parser.parse (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:516:17)
    at Object.parse (REDACTED/d3-scale-cluster/node_modules/webpack/node_modules/acorn/dist/acorn.js:3098:39)
    at Parser.parse (REDACTED/d3-scale-cluster/node_modules/webpack/lib/Parser.js:902:15)
    at NormalModule.<anonymous> (REDACTED/d3-scale-cluster/node_modules/webpack/lib/NormalModule.js:104:16)
    at NormalModule.onModuleBuild (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:310:10)
    at nextLoader (REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:275:25)
    at REDACTED/d3-scale-cluster/node_modules/webpack-core/lib/NormalModuleMixin.js:259:5
    at Storage.finished (REDACTED/d3-scale-cluster/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:38:16)
    at REDACTED/d3-scale-cluster/node_modules/graceful-fs/graceful-fs.js:78:16
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:528:3)
 @ ./~/bindings ^\.\/.*$

ERROR in ./~/bindings/bindings.js
Module not found: Error: Cannot resolve module 'fs' in REDACTED/d3-scale-cluster/node_modules/bindings
 @ ./~/bindings/bindings.js 6:9-22
```

The dist ends up with a bunch of native modules in it, which is no good for the browser.

Can you resolve this issue? I'm guessing maybe we want 2 dist files, one for browser and one for node.

schnerd commented on spec/ScaleCommonSpec.js:
```diff
@@ -0,0 +1,51 @@
+/* globals describe, beforeEach, it, expect */
+var d3scaleCluster = require('../src/index.js');
```
Can you delete ScaleSpec.js since this file seems to replace it?

furstenheim replied:
Yes, missed out in the merge

furstenheim committed:
Fixes pr

furstenheim commented:
So I've fixed the build issue. On one side I've set webpack to ignore ckmeans native, on the other I've added browser field in package json so that if the library is loaded with webpack or browserify it will take the build instead of the index. I've tried it with budo and it seems to work, although I think it would be much nicer to run some tests with karma.
Btw, I wrote an email to sauce labs and they are willing to offer an open source account to make sure that the code runs in different browsers. I won't have time in the next two weeks, but I'll set it up afterwards.

schnerd commented:
@furstenheim merged this, fixed a couple things that weren't working and published as 1.3. Thanks again for the contribution.

schnerd commented:
fyi–i also broke out ckmeans into its own package with all this code: https://github.com/schnerd/ckmeans
